File: src/engine/adapters/sqs/caps.ts
import type { Caps } from '@shared/caps';

// §5.1's sqs row: stream-shaped, batch pagination (poll-on-demand, no addressable position),
// approximate count only, no FK navigation, no console (P10's D13).
export const sqsCaps: Caps = {
  tabular: false,
  documents: false,
  keyValue: false,
  stream: true,
  defaultPageKind: 'stream',
  sql: false,
  // P23 D9: a queue's attributes — visibility timeout, retention, redrive policy, FIFO/dedup,
  // KMS key, ARN — reverses P10's original "no definition" call; nothing in the app showed any of
  // this before. One GetQueueAttributes call, no automatic message read (SPEC §5.1's rule is about
  // ReceiveMessage specifically).
  definition: true,
  // describe() throws E_UNSUPPORTED (sqs/index.ts) — a queue has no column/PK/FK metadata.
  describe: false,
  projection: false,
  serverFilter: false,
  exactCount: false, // ApproximateNumberOfMessages only
  pagination: 'batch',
  foreignKeys: false,
  // sqs/mutate.ts's SendMessage/DeleteMessage land both canInsert and canDelete here. Unlike
  // Kafka, SQS's deleteMessage is a real per-item operation (removes it from the queue via its
  // receipt handle, kept adapter-local — see mutate.ts's own comment). There is still no
  // canUpdate — a delivered message can't be edited in place, only replaced by delete+resend.
  canInsert: true,
  canUpdate: false,
  canDelete: true,
  writable: true,
  transactions: false,
  cancel: true, // the SDK's own abortSignal request option is fully effective (P10's D14)
  fileTransfer: false,
};

File: apps/kira-studio/internal/adapters/sqs/caps.go
package sqs

import (
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
)

// caps is caps.ts's sqsCaps, literally (P58d D18). Stream-shaped, batch pagination (poll-on-
// demand, no addressable position), approximate count only, no console (P10's D13).
var caps = adapters.Caps{
	Tabular:         false,
	Documents:       false,
	KeyValue:        false,
	Stream:          true,
	KeyBrowser:      false,
	DefaultPageKind: page.PageKindStream,
	SQL:             false,
	// P23 D9: a queue's attributes — visibility timeout, retention, redrive policy, FIFO/dedup,
	// KMS key, ARN — reverses P10's original "no definition" call. One GetQueueAttributes call,
	// no automatic message read.
	Definition: true,
	// Describe throws E_UNSUPPORTED (adapter.go) — a queue has no column/PK/FK metadata.
	Describe:     false,
	Projection:   false,
	ServerFilter: false,
	ExactCount:   false, // ApproximateNumberOfMessages only
	Pagination:   adapters.PaginationBatch,
	ForeignKeys:  false,
	// mutate.go's SendMessage/DeleteMessage land both CanInsert and CanDelete here. Unlike Kafka,
	// SQS's DeleteMessage is a real per-item operation (removes it from the queue via its receipt
	// handle, kept adapter-local — see mutate.go's own comment). There is still no CanUpdate — a
	// delivered message can't be edited in place, only replaced by delete+resend.
	CanInsert:    true,
	CanUpdate:    false,
	CanDelete:    true,
	Writable:     true,
	Transactions: false,
	Cancel:       true, // the op's own ctx aborting the in-flight SDK call is fully effective (P58d D3)
	FileTransfer: false,
}
